Sources/NostrSDK/EventCreating.swift

```diff
@@ -225,4 +225,151 @@ public extension EventCreating {
+    ///   - description: A detailed description of the calendar event.
+    ///   - startComponents: An inclusive start date. Must be less than end, if it exists.
+    ///   - endComponents: An exclusive end date. If omitted, the calendar event ends on the same date as start.
+    ///   - location: The location of the calendar event. e.g. address, GPS coordinates, meeting room name, link to video call.
+    ///   - geohash: The [geohash](https://en.wikipedia.org/wiki/Geohash) to associate calendar event with a searchable physical location.
+    ///   - participants: The participants of the calendar event.
+    ///   - hashtags: Hashtags to categorize the calendar event.
+    ///   - references: References / links to web pages, documents, video calls, recorded videos, etc.
+    ///   - keypair: The Keypair to sign with.
+    /// - Returns: The signed ``DateBasedCalendarEvent``.
+    ///
+    /// See [NIP-52](https://github.com/nostr-protocol/nips/blob/master/52.md).
+    func dateBasedCalendarEvent(withName name: String, description: String = "", startComponents: DateComponents, endComponents: DateComponents? = nil, location: String? = nil, geohash: String? = nil, participants: [CalendarEventParticipant]? = nil, hashtags: [String]? = nil, references: [URL]? = nil, signedBy keypair: Keypair) throws -> DateBasedCalendarEvent {
+
+        guard startComponents.isValidDate(in: Calendar(identifier: .iso8601)) else {
+            throw EventCreatingError.invalidInput
+        }
+
+        guard endComponents?.isValidDate(in: Calendar(identifier: .iso8601)) == true else {
+            throw EventCreatingError.invalidInput
+        }
+
+        guard let startDate = startComponents.date else {
+            throw EventCreatingError.invalidInput
+        }
+
+        var tags: [Tag] = []
+
+        // Create an ISO8601 formatter that formats strings in yyyy-mm-dd format so that
+        // the start and end dates conform to NIP-52.
+        let iso8601DateFormatter = ISO8601DateFormatter()
+        iso8601DateFormatter.formatOptions = .withFullDate
+
+        // If the end date is omitted, the calendar event ends on the same date as the start date.
+        if let endDate = endComponents?.date {
+            // The start date must occur before the end date, if it exists.
+            guard startDate < endDate else {
+                throw EventCreatingError.invalidInput
+            }
+
+            tags.append(Tag(name: .unknown("end"), value: iso8601DateFormatter.string(from: endDate)))
+        }
+
+        // Re-arrange tags so that it's easier to read with the identifier and name appearing first in the list of tags,
+        // and the end date being placed next to the start date.
+        tags = [
+            Tag(name: .unknown("d"), value: UUID().uuidString),
+            Tag(name: .unknown("name"), value: name),
+            Tag(name: .unknown("start"), value: iso8601DateFormatter.string(from: startDate))
```

Comment:
I discovered a bug where if `startDate` or `endDate` are in a non-UTC time zone, then what gets computed is actually shifted to be in UTC time. So the intended date in the input could be one day off from what gets added to the tag.

- [ ] Fix time zone issue with date components